### PR TITLE
#ARRUS-145: more flexible sequence, reconstruction, and display programming

### PR DESCRIPTION
**Pull request by PiotrKarwat**
1. BModeDisplay: new functionality (cineloop, persistence, TGC), reconstruction grid taken from reconstruction object.
2. Introduce CustomTxRxSequence (future replacement for SimpleTxRxSequence), seq.txVoltage (future replacement for sys.voltage), rec.sos (speed of sound in reconstruction can be different from speed of sound in tx delays calculation).
3. Make some parameters optional (give them default value or value dependent on other parameters): filterEnable, iqEnable, txPri, tgcStart, tgcSlope, dec, cicOrder.
4. Make some parameters Tx-dependent: txApertureSize, txFocus, txFrequency, txNPeriods, txInvert.
5. Make possible to upload sequence without programming hardware (e.g. for offline reconstruction).

### CODE REVIEW

**Comment by pjarosik** [comment]
@zkLog could you please test it?

**Comment by pjarosik** [comment]
@zkLog ping

File: api/matlab/arrus/BModeDisplay.m [hunk 1]
```diff
@@ -3,38 +3,124 @@
...
+    % 
+    % :param reconstructionObject: object of class "Reconstruction" \
+    %   (obligatory input unless old syntax is used)
+    % :param dynamicRange: two-element vector [min, max], dynamic range \
+    %   limits to apply (optional name-value argument)
+    % :param cineLoopLength: positive scalar, number of frames stored in \
+    %   cineloop (optional name-value argument)
+    % :param persistence: persistence filtration weights (vector) or \
+    %   length (scalar) (optional name-value argument)
+    % :param bmodeTgc: applies linear TGC equal to bmodeTgc[dB]/depth \
```
Comment (zkLog) [inline api/matlab/arrus/BModeDisplay.m:15]: depth units should be given in docstring

File: api/matlab/arrus/BModeDisplay.m [hunk 2]
```diff
@@ -3,38 +3,124 @@
...
+    %   (scalar) (optional name-value argument)
+    % :param bmodeAutoTgcResp: applies linear TGC that adapts in time to \
+    %   the imaging conditions, defines the responsiveness of the \
+    %   adaptation, <0,1> (scalar) (optional name-value argument)
```
Comment (zkLog) [inline api/matlab/arrus/BModeDisplay.m:19]: The 'responsiveness of the adaptation' could be explained more clear.

File: api/matlab/arrus/BModeDisplay.m [hunk 3]
```diff
@@ -3,38 +3,124 @@
     %
```
Comment (zkLog) [inline api/matlab/arrus/BModeDisplay.m:3]: help -> getCineLoop()

**Comment by zkLog** [comment]
Some example script for CustomTxRxSequence would be useful.

**Comment by zkLog** [comment]
When I remove setting  txPri for seqLIN in Us4R_control.m example script and run it, the default txPri value is not present in seqLin object (i.e. seqLin.txPri is 1×0 empty double row vector). The same is for tgcSlope and tgcStart parameters.

**Comment by zkLog** [comment]
Some example script for uploading sequence without programming hardware (e.g. for offline reconstruction) would be useful.

File: api/matlab/arrus/BModeDisplay.m (hunk 1 above)
Comment (PiotrKarwat) [reply api/matlab/arrus/BModeDisplay.m]: corrected: 
    % :param bmodeTgc: applies linear TGC from 0 at top to bmodeTgc[dB] \
    %   at bottom of the image (scalar) (optional name-value argument)

File: api/matlab/arrus/BModeDisplay.m (hunk 2 above)
Comment (PiotrKarwat) [reply api/matlab/arrus/BModeDisplay.m]: "(0-no adaptation, 1-instant adaptation)" added

File: api/matlab/arrus/BModeDisplay.m (hunk 3 above)
Comment (PiotrKarwat) [reply api/matlab/arrus/BModeDisplay.m:3]: Comment added:
    % BModeDisplay class includes a cineloop buffer. The content of the \
    % buffer can be accessed through the getCineLoop method.

**Pull request merged**